onnx/defs/sequence/defs.cc

```diff
@@ -634,4 +634,310 @@ ONNX_OPERATOR_SET_SCHEMA(
+
+    for (int outputIndex = 0; outputIndex < noutputs; outputIndex++) {
+      const auto &body_out_i = body.output(outputIndex);
+      assert(body_out_i.type().has_tensor_type());
+      std::string prefix = MakeString(loopbody_graph_name, "_", body_out_i.name());
+      std::string loopbody_in_name = MakeString(prefix, "_in");
+
+      ValueInfoProto tmp;
+      *tmp.mutable_type()
+          ->mutable_sequence_type()
+          ->mutable_elem_type()
+          ->mutable_tensor_type() = body_out_i.type().tensor_type();
+      tmp.set_name(loopbody_in_name);
+      *loopbody_graph.add_input() = tmp;
+
+      std::string loopbody_out_name = MakeString(prefix, "_out");
+      tmp.set_name(loopbody_out_name);
+      *loopbody_graph.add_output() = tmp;
+
+      NodeProto seq_insert_node;
+      seq_insert_node.set_domain(ONNX_DOMAIN);
+      seq_insert_node.set_op_type("SequenceInsert");
+      seq_insert_node.add_input(loopbody_in_name);
+      seq_insert_node.add_input(body_out_i.name());
+      seq_insert_node.add_output(loopbody_out_name);
+      *loopbody_graph.add_node() = seq_insert_node;
+    }
+  }
+
+  std::vector<FunctionBodyHelper::NodeDef> nodes;
+
+  // TODO: figure out a way to prevent name collisions?
+  auto first_input_name = functionProto.input(0);
+  std::string prefix = MakeString("SequenceMap_", first_input_name);
+  std::string seqlen = MakeString(prefix, "_seqlen");
+  nodes.push_back({{seqlen}, "SequenceLength", {first_input_name}});
+
+  std::string cond_bool = MakeString(prefix, "cond");
+  nodes.push_back(FunctionBodyHelper::Const<bool>(cond_bool, true));
+
+  std::vector<std::string> loop_node_inputs = {seqlen, cond_bool};
+  std::vector<std::string> loop_node_outputs;
+  for (int outputIndex = 0; outputIndex < noutputs; outputIndex++) {
+    auto output_name = functionProto.output(outputIndex);
+    std::string out_prefix = MakeString("SequenceMap_", output_name);
+
+    std::string seqempty_name = MakeString(out_prefix, "_seqempty");
+    int64_t dtype = g_outputs[outputIndex].type().tensor_type().elem_type();
+    nodes.push_back({
+      {seqempty_name},
+      "SequenceEmpty",
+      {},
+      {MakeAttribute("dtype", dtype)}
+    });
+    loop_node_inputs.push_back(seqempty_name);
+    loop_node_outputs.push_back(output_name);
+  }
+
+  nodes.push_back({
+    loop_node_outputs,
+    "Loop",
+    loop_node_inputs,
+    {MakeAttribute("body", loopbody_graph)}
+  });
+
+  auto func_nodes = FunctionBodyHelper::BuildNodes(nodes);
+  for (const auto& node : func_nodes) {
+    auto new_node = functionProto.add_node();
+    new_node->CopyFrom(node);
+  }
+
+  return true;
+}
+
+ONNX_OPERATOR_SET_SCHEMA(
+    SequenceMap,
+    16,
+    OpSchema()
+        .SetDoc(SequenceMap_ver16_doc)
+        .Attr(
+            "body",
+            "The graph run for each sample in the sequence(s). "
+            "It should have as many tensor inputs and tensor outputs as inputs and "
```

Comment:
Change `tensor inputs/outputs` to plain `inputs/outputs`? Makes it somewhat future proof, since conceptually there is no reason why it should be tensors (even though the type signature below constrains it currently).

Reply:
Goop point. Done.